Proposed 1.12.0 (#4690)
XRPLF/rippled: RELEASENOTES.md

- **`fixReducedOffersV1`**: Reduces the occurrence of order books that are blocked by reduced offers. [#4512](https://github.com/XRPLF/rippled/pull/4512)

- Added binary hardening and linker flags to enhance security during the build process. [#4603](https://github.com/XRPLF/rippled/pull/4603)

- Updated build dependencies to the most recent versions in Conan Center. [#4595](https://github.com/XRPLF/rippled/pull/4595)

- Updated Conan recipe for NuDB. [#4615](https://github.com/XRPLF/rippled/pull/4615)

- Added a pre-commit hook that runs the clang-format linter locally before committing changes. To install this feature, see: [CONTRIBUTING](https://github.com/XRPLF/xrpl-dev-portal/blob/master/CONTRIBUTING.md). [#4599](https://github.com/XRPLF/rippled/pull/4599)

- Added quality-of-life improvements to workflows, using new [concurrency control](https://docs.github.com/en/actions/using-jobs/using-concurrency) features. [#4597](https://github.com/XRPLF/rippled/pull/4597)

> - Added an Artifactory to the `nix` workflow to improve build times. [#4556](https://github.com/XRPLF/rippled/pull/4556)

Reviewer: My previous comment was mostly about the fact that these items are less important than the actual "new features" that people might want to read about. Since they're all build system improvements, they don't really belong in the category with top billing. Maybe those should be lumped in with Docs (a "Docs and Build System" category) or in their own category down below?

Also, alluded to in the other sections but not actually called out here is that this release makes API v2 available. That should probably get a pretty big mention under "New Features"

Reviewer 2: We decided not to make API v2 available yet. That will be for the next release (maybe 1.13.0, maybe 2.0.0, ...)

Reviewer 2: Given API v2 is still in beta, I will remove the bullets for API v2 changes. Readers who care can still see the "Full Commit Log" - that should be sufficient.

- Added WebSocket and RPC port info to `server_info` responses. [#4427](https://github.com/XRPLF/rippled/pull/4427)

### Bug Fixes and Performance Improvements

- Fixed an incorrect error response when there are missing fields in the API v2 `ledger_entry` method. [#4552](https://github.com/XRPLF/rippled/pull/4552)

- Updated checkout versions to resolve warnings during Github jobs. [#4598](https://github.com/XRPLF/rippled/pull/4598)

- Added an error response to the API v2 `account_info` method when you include an invalid `signer_lists` value. [#4585](https://github.com/XRPLF/rippled/pull/4585)

- Fixed an issue with the debug package build. [#4591](https://github.com/XRPLF/rippled/pull/4591)

- Added additional error responses to the API v2 `AccountTx` method. [#4571](https://github.com/XRPLF/rippled/pull/4571)

- Fixed build references to deleted `ServerHandlerImp`. [#4592](https://github.com/XRPLF/rippled/pull/4592)

- Fixed package definitions for Conan. [#4485](https://github.com/XRPLF/rippled/pull/4485)

> - Changed the output type of the `mulDiv()` function from `std::pair<bool, uint64_t>` to `std::optional`. [#4243](https://github.com/XRPLF/rippled/pull/4243)

Reviewer: Some context from the issue that may be helpful to call out: the goal of this is to make it more straightforward to catch and handle overflow errors.

- Updated `Handler::Condition` enum values to make the code less brittle. [#4239](https://github.com/XRPLF/rippled/pull/4239)

> - Renamed `ServerHandlerImp` to `ServerHandler`. [#4516](https://github.com/XRPLF/rippled/pull/4516)

Reviewer: Add the reference to 4592 here and remove its own bullet point from line 72

```suggestion
- Renamed `ServerHandlerImp` to `ServerHandler`. [#4516](https://github.com/XRPLF/rippled/pull/4516), [#4592](https://github.com/XRPLF/rippled/pull/4592)
```

> - Removed the deprecated `accepted`, `seq`, `hash`, and `totalCoins` fields from the `ledger` method. [#4244](https://github.com/XRPLF/rippled/pull/4244)

Reviewer: This might be worth moving up to "New Features"? It's an API change at least, although a pretty minor one.

Also, the linked ticket description is slightly wrong; there is no `seq` field to be removed; it's `seqNum`.

 
Have new ideas? Need help with setting up your node? [Please open an issue here](https://github.com/xrplf/rippled/issues/new/choose).

# Introducing XRP Ledger version 1.12.0

Version 1.12.0 of `rippled`, the reference server implementation of the XRP Ledger protocol, is now available. This release adds new features and bug fixes, and introduces these amendments:

> - `featureAMM`

Reviewer 2: I'm not sure if we should include the `feature` prefix. For [feature](https://xrpl.org/feature.html), when specifying the name in the `feature` field, should the `feature` prefix be included?

Reviewer: The `feature` prefix is _not_ part of the amendment name.

```suggestion
- `AMM`
```

> - `featureClawback`

Reviewer 2: ```suggestion
- `Clawback`
```

- `fixReducedOffersV1`

[Sign Up for Future Release Announcements](https://groups.google.com/g/ripple-server)

<!-- BREAK -->

## Action Required

Three new amendments are now open for voting according to the XRP Ledger's [amendment process](https://xrpl.org/amendments.html), which enables protocol changes following two weeks of >80% support from trusted validators.

> If you operate an XRP Ledger server, upgrade to version 1.12.0 by September 20 2023 to ensure service continuity. The exact time that protocol changes take effect depends on the voting decisions of the decentralized network.

Reviewer 2: ```suggestion
If you operate an XRP Ledger server, upgrade to version 1.12.0 by September 20, 2023 to ensure service continuity. The exact time that protocol changes take effect depends on the voting decisions of the decentralized network.
```

> If you operate an XRP Ledger server, upgrade to version 1.12.0 by September ... to ensure service continuity. The exact time that protocol changes take effect depends on the voting decisions of the decentralized network.

Reviewer 2: ```suggestion
If you operate an XRP Ledger server, upgrade to version 1.12.0 by September 20, 2023 to ensure service continuity. The exact time that protocol changes take effect depends on the voting decisions of the decentralized network.
```

## Install / Upgrade

On supported platforms, see the [instructions on installing or updating `rippled`](https://xrpl.org/install-rippled.html).

## New Amendments

> - **`featureAMM`**: Introduces an automated market maker (AMM) protocol to the XRP Ledger's decentralized exchange, enabling you to trade assets without a counterparty. For more information about AMMs, see: [link](). [#4294](https://github.com/XRPLF/rippled/pull/4294)

Reviewer 2: ```suggestion
- **`featureAMM`**: Introduces an automated market maker (AMM) protocol to the XRP Ledger's decentralized exchange, enabling you to trade assets without a counterparty. For more information about AMMs, see: [Automated Market Maker](https://opensource.ripple.com/docs/xls-30d-amm/amm-uc/). [#4294](https://github.com/XRPLF/rippled/pull/4294)
```

Reviewer: Drop the `feature` prefix here and on the next couple lines as well

- **`Clawback`**: Allows issuers to add the `lsfAllowTrustLineClawback` flag to an issuing account. This enables the account to recover, or _claw back_, issued tokens after they're distributed to accounts. For additional documentation on this feature, see: [#4553](https://github.com/XRPLF/rippled/pull/4553).

> - **`fixReducedOffersV1`**: Reduces the occurence of order books that are blocked by reduced offers. [#4512](https://github.com/XRPLF/rippled/pull/4512)

Reviewer 2: ```suggestion
- **`fixReducedOffersV1`**: Reduces the occurrence of order books that are blocked by reduced offers. [#4512](https://github.com/XRPLF/rippled/pull/4512)
```

## Changelog

> ### New Features and Improvements

Reviewer: I think this division of categories is not very useful. I'd like to go back to something more like [v1.9.1](https://github.com/XRPLF/rippled/releases/tag/1.9.1) when "New Features and Amendments" were one item—a list of things people might actually find useful about the release—and "Bug Fixes and Improvements"—a list of things that changed for the better but users don't need to do anything for—is separate.

As it stands, most of the bullet points in this section (everything except the WebSocket RPC port one) are improvements to the code, build system, and other behind-the-scenes stuff that someone just using the server doesn't need to care about at all.
